### PR TITLE
Fixed memory leak in nfc_initiator_select_passive_target()

### PR DESCRIPTION
**Pull request by martin-vitek**

### CODE REVIEW

File: libnfc/nfc.c [hunk 1]
```diff
@@ -579,6 +579,7 @@ nfc_initiator_select_passive_target(nfc_device *pnd,
     free(abtTmpInit);
   } else if (nm.nmt == NMT_ISO14443A) {
     abtInit = abtTmpInit;
+    free(abtTmpInit);
     iso14443_cascade_uid(pbtInitData, szInitData, abtInit, &szInit);
```
Comment (smortex) [inline libnfc/nfc.c:583]: Here is now a use-after-free for abtInit (which is assigned the address of abtTmpInit which is freeed).

File: libnfc/nfc.c [hunk 2]
```diff
@@ -587,8 +588,6 @@ nfc_initiator_select_passive_target(nfc_device *pnd,
     szInit = szInitData;
   }
   HAL(initiator_select_passive_target, pnd, nm, abtInit, szInit, pnt);
-
-  free(abtTmpInit);
```
Comment (smortex) [inline libnfc/nfc.c:591]: What are you trying to fix?  A double-free error?  If so, I feel like this should be the only free() after `nfc_device_validate_modulation()` returns successfuly.

**Review by smortex — COMMENTED** [review]
This code looks convoluted, maybe I read it badly but I do not thing this fix improve the situation.

**Comment by martin-vitek** [comment]
I checked my app, which is using libnfc, with Valgrind and it discovered, that function `nfc_initiator_select_passive_target()` is leaking 12B.

https://github.com/nfc-tools/libnfc/blob/d9a04a54ffd99348cbbbf7c2e5e4747c2b8fe404/libnfc/nfc.c#L576-L591

Problem is in calling macro `HAL` before free(), because the macro is returning from `nfc_initiator_select_passive_target()`, so the free() afterwars is never called. And in case `nm.nmt == NMT_ISO14443A`, `abtTmpInit` isn't deallocated.

But as I'm looking at it now, my solution isn't good, because `abtInit` is used after my added free() by `iso14443_cascade_uid()`. And it is used as `abtInit`, which is pointing to deallocated address, also by macro `HAL` and `initiator_select_passive_target`.

**Comment by smortex** [comment]
Ah, I see…  thanks you for the context, I did not checked in that area.  Returning from a macro is never a good idea…  We can either do some big refactoring to avoid this macro completely, or a short-term easier fix would be to call the macro from another internal function, save the returned value, free memory and return the saved value.

**Comment by martin-vitek** [comment]
Yeah, reworking the `HAL` macro would be a good idea, but as a much easier fix, I simply replaced `malloc()` with `alloca()`, so the memory is allocated on the stack and automatically deallocated  when exiting the function. As the size of the allocated memory is small, use of `alloca()` should be fine.

**Comment by smortex** [comment]
Hum :thinking: I am concerned about portability here… I don't know if many people use the libnfc code in embedded systems, but they might not have an `alloca()` function available at hand.  @nfc-tools/core  though?

File: libnfc/nfc.c (hunk 1 above)
Comment (tungnduong) [reply libnfc/nfc.c:581]: Agree with @smortex , this fix doesn't make sense.  abtInit pointing to the freed memory location, and is subsequently passed to iso14443_cascade_uid.

**Pull request closed without merging**